artemis-server/src/main/java/org/apache/activemq/artemis/core/server/federation/FederatedQueueConsumerImpl.java

```diff
@@ -155,6 +167,67 @@ private synchronized void connect() throws Exception {
       }
    }
 
+   interface QueueHandle {
+      long getMessageCount();
+      int getCreditWindow();
+   }
+
+   private QueueHandle createQueueHandle(ActiveMQServer server, ClientSession.QueueQuery queryResult) {
+      final Queue queue = server.locateQueue(queryResult.getName());
+      int creditWindow = DEFAULT_CONSUMER_WINDOW_SIZE;
+
+      final Integer defaultConsumerWindowSize = queryResult.getDefaultConsumerWindowSize();
+      if (defaultConsumerWindowSize != null) {
+         creditWindow = defaultConsumerWindowSize.intValue();
+         if (creditWindow <= 0) {
+            creditWindow = DEFAULT_CONSUMER_WINDOW_SIZE;
+            logger.trace("{} override non positive queue consumerWindowSize with {}.", this, creditWindow);
+         }
+      }
+
+      final int finalCreditWindow = creditWindow;
+      return new QueueHandle() {
+         @Override
+         public long getMessageCount() {
+            return queue.getMessageCountForRing();
+         }
+
+         @Override
+         public int getCreditWindow() {
+            return finalCreditWindow;
+         }
+      };
+   }
+
+   private void scheduleCreditOnEmpty(final int delay, final QueueHandle handle) {
+      if (handle != null) {
```

Comment:
Personally I'd treat a null handle being passed here as a terminal event and throw since it's probably gonna break something if you try and schedule for credit and don't actually schedule anything to replenish credit.

Reply:
agree, that check was just defensive but unnecessary, the consumer can go stale with a pending check but that is already covered and any failure will result in the session/consumer getting recreated, so all that in necessary is that we don't reschedule in that case.
thanks for the feedback.